### PR TITLE
Make Analysis providers possible to give back data points

### PR DESCRIPTION
**Pull request by nakabonne**
**What this PR does / why we need it**:
To allow analysis dynamically (https://github.com/pipe-cd/pipe/pull/2385), analysis providers have to give back data points to compare. This PR adds a new method named `SelectPoints`.

Each `SelectPoints` implementation is almost the same as `Evaluate` already implemented and works stable.

I'm looking to update the static analyzer as well to adopt this way later.

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/piped/analysisprovider/metrics/provider.go [hunk 1]
```diff
@@ -28,11 +28,22 @@ var (
...
 	// Evaluate runs the given query against the metrics provider,
 	// and then checks if the results are expected or not.
 	// Returns the result reason if non-error occurred.
 	// The first value "expected" must be false if err isn't nil.
+	// TODO: Do not evaluate data points by Analysis providers
+	//   Instead, the executor should do that by using SelectPoints().
 	Evaluate(ctx context.Context, query string, queryRange QueryRange, evaluator Evaluator) (expected bool, reason string, err error)
+
+	// SelectPoints gives back data points within the given range.
+	SelectPoints(ctx context.Context, query string, queryRange QueryRange) (points []DataPoint, err error)
```
Comment (nghialv) [inline pkg/app/piped/analysisprovider/metrics/provider.go:41]: nit: `QueryPoints` and `queryRange` -> `range`

**Comment by nghialv** [comment]
Great work!
Just left a nit.

File: pkg/app/piped/analysisprovider/metrics/provider.go (hunk 1 above)
Comment (nakabonne) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: Thanks. Does it mean `type QueryRange struct` should be `Range`?

Comment (nakabonne) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: Ah, you mean it should be `QueryPoints ` instead of `SelectPoints`.

Comment (nghialv) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: I was talking about the name, not the type. w

Comment (nghialv) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: `SelectPoints` -> `QueryPoints`
`queryRange` -> `range`

Comment (nakabonne) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: Honestly, the word `range` isn't available in Go because of the reserved word.

Comment (nakabonne) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: Alright 👍

Comment (nghialv) [reply pkg/app/piped/analysisprovider/metrics/provider.go:41]: Ah right. So that is fine. 👍

**Comment by khanhtc1202** [comment]
Way to go
/approve

**Pull request merged**